packages/beacon-node/src/chain/validation/voluntaryExit.ts

```diff
@@ -1,58 +1,250 @@
-  return validateVoluntaryExit(chain, voluntaryExit);
-}
+  const result = await validateVoluntaryExitDetailed(chain, voluntaryExit);
 
-async function validateVoluntaryExit(
-  chain: IBeaconChain,
-  voluntaryExit: phase0.SignedVoluntaryExit,
-  prioritizeBls = false
-): Promise<void> {
-  // [IGNORE] The voluntary exit is the first valid voluntary exit received for the validator with index
-  // signed_voluntary_exit.message.validator_index.
-  if (chain.opPool.hasSeenVoluntaryExit(voluntaryExit.message.validatorIndex)) {
-    throw new VoluntaryExitError(GossipAction.IGNORE, {
-      code: VoluntaryExitErrorCode.ALREADY_EXISTS,
+  if (!result.isValid) {
+    //biome-ignore lint/style/noNonNullAssertion: error is guaranteed to exist when isValid is false
+    throw new VoluntaryExitError(result.error!.action, {
+       //biome-ignore lint/style/noNonNullAssertion: error is guaranteed to exist when isValid is false
+            code: result.error!.code,
     });
   }
+}
 
-  // What state should the voluntaryExit validate against?
-  //
-  // The only condition that is time sensitive and may require a non-head state is
-  // -> Validator is active && validator has not initiated exit
-  // The voluntaryExit.epoch must be in the past but the validator's status may change in recent epochs.
-  // We dial the head state to the current epoch to get the current status of the validator. This is
-  // relevant on periods of many skipped slots.
+/**
+ * Process pending voluntary exits at each epoch
+ * Should be called by the beacon chain on epoch transitions
+ */
+export async function processPendingVoluntaryExits(
+  chain: IBeaconChain,
+  pendingPool: PendingVoluntaryExitPool,
+  network: { publishVoluntaryExit(exit: phase0.SignedVoluntaryExit): Promise<void> }
+): Promise<void> {
   const state = await chain.getHeadStateAtCurrentEpoch(RegenCaller.validateGossipVoluntaryExit);
+  const currentEpoch = state.epochCtx.epoch;
 
-  // [REJECT] All of the conditions within process_voluntary_exit pass validation.
-  // verifySignature = false, verified in batch below
-  if (!isValidVoluntaryExit(chain.config.getForkSeq(state.slot), state, voluntaryExit, false)) {
-    throw new VoluntaryExitError(GossipAction.REJECT, {
-      code: VoluntaryExitErrorCode.INVALID,
-    });
+  const toRemove: number[] = [];
+
+  for (const [validatorIndex, cached] of pendingPool.getAllPending()) {
+
+    if (cached.lastCheckedEpoch === currentEpoch) {
+      continue;
+    }
+
+    try {
+      const result = await validateVoluntaryExitDetailed(chain, cached.exit);
+
+      if (result.isValid) {
+        chain.opPool.insertVoluntaryExit(cached.exit);
+        await network.publishVoluntaryExit(cached.exit);
+        toRemove.push(validatorIndex);
+      } else if (!result.error?.isTransient) {
+        toRemove.push(validatorIndex);
+      }
+    // biome-ignore lint/correctness/noUnusedVariables: error variable unused but required by catch syntax
+    } catch (e) {
```

Comment:
none of these `biome-ignore` are necessary

eg. you can just do

```suggestion
    } catch {
```